Monitoring & Learning Quiz creation and json api exposure

imsickofmaps opened the pull request:
## Basics
- Needs a M&L quiz model creating first (this is not the same as weekly quiz)
- Each quiz will have a name and 4 questions with multiple options. 
- Total quiz question and answers should not exceed 160 characters (this should be validated in admin UI - server side) per question
- Admin area should show overview information for the list display 
- A quiz can only be set to active if there are four questions.
## API

There need to be three endpoints

**GET /api/mandl/all/**

Returns an array of quiz ids (this unique id for the quiz in the DB)

```
{"quizzes":[3, 4, 5]}
```

**GET /api/mandl/**

Returns a dict of the active quizzes in the same format as the individual one below.

```
{
"quizzes": {
    "mandl_quiz_1": {"start":"q1", "etc": "etc"},
    "mandl_quiz_2": {"start":"q1", "etc": "etc"}
}
```

**GET /api/mandl/{id}**

```
{
"quiz": {
   "start": "q_1",
   "questions": {
    "q_1": {
          "question": "Is this fake question one?",
          "choices": [
              [
                  "q_2",
                  "Yes"
              ],
              [
                  "q_2",
                  "No"
              ]
          ]
      },
      "q_2": {
          "question": "Is this fake question two?",
          "choices": [
              [
                  "main_menu",
                  "Of course"
              ],
              [
                  "main_menu",
                  "No way!"
              ]
          ]
      }          
  }
}
```

smn commented:
It may be useful to already specify as part of the options if it's a correct or an incorrect answer. It could possibly make the state juggling in the Javascript app a bit easier.

imsickofmaps commented:
There's no correct/incorrect for the M&L quiz. Its more of a profiling survey to be honest from what I can tell. I will do that for the real "Weekly quiz".

smn commented:
/me swallows comment.

gitumarkk committed:
Created Monitor and Learning Model

gitumarkk committed:
Registered M&L with admin

gitumarkk committed:
Did a south migration with M & L

gitumarkk committed:
Changed Model Class Names

gitumarkk committed:
Created reference unicode for QuizId

gitumarkk committed:
Created view for ques and answers on same page

Mike, please have a look at this solution on the admin and see if I am
on the right ladder (I am actually impressed at Django for its
capabilities).

gitumarkk committed:
correcting character buffer Type Error

Can't add unicode without it affecting the code, so skipping it for now.

gitumarkk committed:
check to ensure total character length < 160

gitumarkk committed:
Char length check < 160 working (BOOM)

gitumarkk committed:
Giving Quiz_id a human readable name

It breaks with __unicode__(), but seems to work with __str__()

gitumarkk committed:
Moved M&L to a new app

gitumarkk committed:
Limit questions asked to 4

There is a bug in the code where if the Quiz Id is not provided it
breaks, will sort it out.

gitumarkk commented:
@imsickofmaps made commit for code

gitumarkk committed:
Completed model and admin section

gitumarkk committed:
Adding comments to code

gitumarkk committed:
Hiding completed field in quiz id admin

gitumarkk commented:
@imsickofmaps made final(ish) changes to monitor and learning Model, starting on API

gitumarkk committed:
Creating Tests for M&L API (TDD)

gitumarkk committed:
Adding url of API to urls.py

gitumarkk committed:
Added mandl quiz id API

gitumarkk committed:
Deleted a document committed by mistake

gitumarkk committed:
Switching to articles changes

gitumarkk committed:
Working on M&L API on hold

gitumarkk committed:
Undoing mistake in settings.py

gitumarkk committed:
Coding the M&L API to respond to with quiz ID

gitumarkk committed:
Adding tests for M&L API

gitumarkk commented:
@imsickofmaps made final commit for M & L

imsickofmaps committed:
Merge branch 'develop' into feature/issue-9-monitoring-and-learning-quiz-creation-and-json-api-exposure

Fixed conflicts:
	articles/admin.py
	ghr/urls.py

gitumarkk committed:
Created initial structure & code for opinions

gitumarkk committed:
Revert "Created initial structure & code for opinions"

This reverts commit da3dcc70a28eb8dd19302fed14dc6257587ddbf0.

gitumarkk committed:
Fixing bug (initially assumed objects are sorted)

Fixed a bug where initially I assumed all objects are sorted from the
database (was testing with sqlite3), when tested with postgresql, the
query was sorted differently hence made a fix that assumes that data is
unsorted.

gitumarkk committed:
Added comments to code

gitumarkk committed:
made pep8 changes

gitumarkk committed:
Added name field to Model and did Schemamigration

gitumarkk committed:
Removed limit of choosing the latest quiz

User can add questions to any uncompleted quiz.

gitumarkk committed:
changed answers validation and added pep8 changes

gitumarkk committed:
Added validation on # of questions & answers

gitumarkk commented:
@imsickofmaps made commit for questions and answers validation.

imsickofmaps commented on monitor_and_learn/admin.py:
```diff
@@ -0,0 +1,120 @@
+        # also checks if the answer has been submitted and Foreign key has
+        # been included
+        super(MonitorAndLearningQuizAnswerFormset, self).clean()
+        i = 0
```
I don't mind us using `i` for genuine iterators - but can we use descriptive names for things like this to make core more readable? I wouldn't have called this `answers`.

imsickofmaps commented on monitor_and_learn/admin.py:
```diff
@@ -0,0 +1,120 @@
+class MonitorAndLearningQuizAnswerInline(admin.StackedInline):
+    # This class sets the answers to be on the same page as the parent question
+    model = MonitorAndLearningQuizAnswer
+    extra = 2
```
This is where comments should be... you know what `extra` does right now. But will you remember? Or will the schmuck who has to edit this in 6 month?

imsickofmaps commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,89 @@
+            # Need to store end point for main_menu
+            max_id.append(quiz_ids[i].data["id"])
+
+        max_id = max(max_id)
```
Please don't reuse variables with a different type. While it works, it just leaves you open to issues. A cleaner way to do this is something like `last_question_id = max([quiz_ids[i].data["id"] for i in quiz_ids])`

imsickofmaps commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,89 @@
+        if isinstance(data_dict, dict):
+            if 'meta' in data_dict:
+                del(data_dict['meta'])
+            a = []
```
Use descriptive variables

imsickofmaps commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,89 @@
+            # Looping through data and adding it to dict_item for final output
+            q_id = "q_%s" % quiz_ids[i].data["id"]
+            dict_item[q_id] = {"choices": []}
+            x = []
```
Use descriptive variables

imsickofmaps commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,89 @@
+                # For loop adds answer to a list that will be appended to dict.
+                # also adds go to variables to the answer
+                if quiz_ids[i].data["id"] == max_id:
+                    y = "main_menu"
```
Use descriptive variables

imsickofmaps commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,89 @@
+                z = quiz_ids[i].data["quiz_ids"][j].data["answer"]
+                x.append([y, z])
+
+            dict_item[q_id] = {"question": quiz_ids[i].data["question"],
```
Use descriptive variables

gitumarkk committed:
Renaming variable names to make them logical

gitumarkk committed:
Adding comments to code

gitumarkk committed:
Using generator for last_question_id not for loop

gitumarkk committed:
pep8 changes

gitumarkk committed:
Deleted unused variables

gitumarkk committed:
Addes comments to admin and delete print statment

gitumarkk committed:
added code for /mandl/all/

gitumarkk committed:
Added unittests for /mandl/all/

gitumarkk committed:
Adding active filter to API

gitumarkk committed:
Testing for active filter

gitumarkk commented:
@imsickofmaps due to the internal workings of tastypie where it is much easier to access the individual foreign key data from inbuilt methods I have changed the url path as follows:

**GET /api/mandl/all/ returns**
`
{"quizzes":[3, 4, 5]}
`
**GET /api/mandl/ returns**
`
{
"quizzes": {
    "mandl_quiz_1": {"start":"q1", "etc": "etc"},
    "mandl_quiz_2": {"start":"q1", "etc": "etc"}
}
`

imsickofmaps commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,148 @@
+    def get_quiz_ids(self, request, **kwargs):
+        quizzes = super(MonitorAndLearningQuizIDResource, self).get_object_list(request)
+        quizzes = self.get_object_list(request)
+        quiz_id = []
```
This should be `quiz_ids`

imsickofmaps commented:
@smn can you review this when you have a moment?

gitumarkk committed:
changing reused variable names

gitumarkk committed:
deleting test.py stub file

smn commented on monitor_and_learn/admin.py:
```diff
@@ -0,0 +1,139 @@
+                for item in range(len(query)):
+                    choices.append((query[item].pk, query[item].name))
+
+        self.fields['quiz_id'].choices = choices
```
Can we chat in #prk-dev a bit about what's happening here? I'm not sure I'm following and seems a bit too complex.

gitumarkk replied:
@smn basically I am trying to edit what Quizzes get shown in the Questions table so that if there are 4 questions (max allowed per quiz) the quiz Ids is automatically hidden, however if there are less than four questions the current Questions->Quiz Id Foreign Key field defaults to the last Quiz Id.

smn replied:
This whole `choices` thing still seems a bit overcomplex. I'll leave it for now & merge this stuff, we'll probably visit again at a later stage.

smn commented on monitor_and_learn/admin.py:
```diff
@@ -0,0 +1,139 @@
+        # an DoesNotExist exception, despite the validation been active
+            query = (MonitorAndLearningQuizQuestion.objects.all().
+                     filter(quiz_id=self.instance.quiz_id))
+            if len(query) >= 3:
```
`query.count()`

smn commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,148 @@
+        """
+        if isinstance(data_dict, dict):
+            if 'meta' in data_dict:
+                del(data_dict['meta'])
```
Don't need to change but this is handy to know:

``` python
>>> d = {'foo': 'bar'}
>>> d.pop('boo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'boo'
>>> d.pop('boo', None)
>>> d.pop('foo', None)
'bar'
>>> d
{}
>>>
```

smn commented on monitor_and_learn/api.py:
```diff
@@ -0,0 +1,148 @@
+                quizzes_dict[mandl_id] = {"questions": questions_dict,
+                                          "start": "q_%s" % first_question_id}
+            data_dict['quizzes'] = quizzes_dict  # Adding the quiz ids to quizzes
+            del (data_dict['objects'])
```
although it doesn't hurt, `del` is a statement, you don't need the brackets.

smn commented on monitor_and_learn/tests.py:
```diff
@@ -0,0 +1,16 @@
+This file demonstrates writing tests using the unittest module. These will pass
+when you run "manage.py test".
+
+Replace this with more appropriate tests for your application.
```
Or just remove the file if you're not doing any tests in this file.

smn commented on monitor_and_learn/tests/test_api.py:
```diff
@@ -0,0 +1,126 @@
+                                ["main_menu", "A1Dc"]]))
+        # print json_item
+
+    def test_invalid_addres(self):
```
typo

smn commented on monitor_and_learn/tests/test_api.py:
```diff
@@ -0,0 +1,126 @@
+                      'api_name': 'api', "pk": 1})
+        response = self.client.get(url)
+        json_item = json.loads(response.content)
+        self.assertEqual(response.request["PATH_INFO"], "/api/mandl/1/")
```
I'd consider these tests unnecessary, the actual url this resource sits on isn't imported and should be changeable without the test failing.

gitumarkk committed:
Made changes suggested by smn

gitumarkk committed:
Changed function that returns unicode

gitumarkk committed:
Set the completed field to editable=False

gitumarkk committed:
Used query.count() instead of len(query)

gitumarkk committed:
Removed () from del()

gitumarkk committed:
changed quiz_id to quiz_ids

gitumarkk committed:
removing unnecessary tests

gitumarkk committed:
fixing typo in tests

smn commented:
I've got a test failure in this, but that exists in `develop` as well. I'll land this but we'll need to fix the [failing test](https://travis-ci.org/praekelt/go-nike-ghr/jobs/8966249).

gitumarkk commented:
@smn The test fails due to the fixture being outdated i.e. the tests checks for the article published last week, however the fixture was written 1.5 weeks ago. Was wondering if it was okay to use fixtures for this particular test or rather creating the article model in `def setUp()` would be better
